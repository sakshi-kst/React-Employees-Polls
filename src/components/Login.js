import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Segment, Grid, Header, Form, Loader, Dimmer } from 'semantic-ui-react';
import { setAuthUser } from '../actions/authUser';

export class Login extends Component {
  state = {
    loading: false
  };
  handleLoading = () => {
    this.setState({ loading: true });
  };

  render() {
    return (
      <Fragment>
        <Segment.Group>
          <LoginHeader />
          <LoginGridLayout
            form = {<ConnectedLoginForm onLoading = {this.handleLoading} />}
            loading = {this.state.loading}
          />
        </Segment.Group>
      </Fragment>
    );
  }
}

const LoginHeader = () => (
  <Header as = "h4" block attached = "top" textAlign = "center">
    <Header.Content>Employee Polls Web App</Header.Content>
    <Header.Subheader>Please Sign In to continue</Header.Subheader>
  </Header>
);

const LoginGridLayout = ({ form, loading }) => (
  <div>
    <Grid padded textAlign = "center">
      <Grid.Row className = "login">
        <Grid.Column width = {16}>
          {loading === true && (
            <Dimmer active inverted>
              <Loader inverted content = "Loading" />
            </Dimmer>
          )}
          {form}
        </Grid.Column>
      </Grid.Row>
    </Grid>
  </div>
);

class LoginForm extends Component {
  static propTypes = {
    onLoading: PropTypes.func.isRequired
  };

  state = {
    value: ''
  };

  onChange = (e, {value}) => {
    this.setState({value});
  };
  
  handleSubmit = e => {
    e.preventDefault();
    const { onLoading, setAuthUser } = this.props;
    const authUser = this.state.value;

    new Promise((resolve, reject) => {
      onLoading();
      setTimeout(() => resolve(), 500);
    }).then(() => setAuthUser(authUser));
  };

  generateDropdownData = () => {
    const { users } = this.props;

    return users.map(user => ({
      key: user.id,
      text: user.name,
      value: user.id,
      image: { avatar: true, src: user.avatarURL }
    }));
  };
  
  render() {
    const { value } = this.state;
    const disabled = value === '' ? true : false;

    return (
      <Form onSubmit = {this.handleSubmit}>
        <div data-testid = 'Test Component'>
          <Header as = "h2" color = "green">
            Sign In
          </Header>
          <Form.Dropdown
            placeholder = "Select an Employee"
            fluid
            selection
            scrolling
            options = {this.generateDropdownData()}
            value = {value}
            onChange = {this.onChange}
            data-testid = 'dropdown'
            required
          />
          <Form.Button content = "Login" positive fluid disabled = {disabled} data-testid = 'button' />
        </div>
      </Form>
    );
  }
}

function mapStateToProps({users}) {
  return {
    users: Object.values(users)
  };
}

const ConnectedLoginForm = connect(mapStateToProps, {setAuthUser})(LoginForm);

export default Login;