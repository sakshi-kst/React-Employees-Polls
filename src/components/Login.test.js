import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { within } from '@testing-library/dom'
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Login from './Login';

const mockStore = configureStore([thunk]);

describe('Login', () => {
    let store, component;

    beforeEach(() => {
        store = mockStore({
            users: {
                sarahedo: {
                    id: 'sarahedo',
                    name: 'Sarah Edo',
                },
                tylermcginnis: {
                    id: 'tylermcginnis',
                    name: 'Tyler McGinnis',
                },
                mtsamis: {
                    id: 'mtsamis',
                    name: 'Mike Tsamis',
                },
                zoshikanlu: {
                    id: 'zoshikanlu',
                    name: 'Zenobia Oshikanlu',
                },
            },
        });
    });

    it('will match snapshot', () => {
        component = render(
            <Provider store = {store}>
                <Login />
            </Provider>
        );

        expect(component).toMatchSnapshot();
    });

    it('should select a user from the dropdown', () => {
        const { getByTestId } = render(
            <Provider store = {store}>
                <Login />
            </Provider>
        )

        component = getByTestId('Test Component');
        const dropdownOptions = within(component).getAllByRole('option');
        const option = dropdownOptions[1];
        
        fireEvent.click(option);
        expect(option.getAttribute('aria-selected')).toBe('true');
      });
})