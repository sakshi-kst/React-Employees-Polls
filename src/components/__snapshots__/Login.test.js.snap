// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui segments"
      >
        <h4
          class="ui block top attached center aligned header"
        >
          <div
            class="content"
          >
            Employee Polls Web App
          </div>
          <div
            class="sub header"
          >
            Please Sign In to continue
          </div>
        </h4>
        <div>
          <div
            class="ui padded center aligned grid"
          >
            <div
              class="row login"
            >
              <div
                class="sixteen wide column"
              >
                <form
                  class="ui form"
                >
                  <div
                    data-testid="Test Component"
                  >
                    <h2
                      class="ui green header"
                    >
                      Sign In
                    </h2>
                    <div
                      class="required field"
                    >
                      <div
                        aria-expanded="false"
                        class="ui fluid selection scrolling dropdown"
                        data-testid="dropdown"
                        required=""
                        role="listbox"
                        tabindex="0"
                      >
                        <div
                          aria-atomic="true"
                          aria-live="polite"
                          class="divider default text"
                          role="alert"
                        >
                          Select an Employee
                        </div>
                        <i
                          aria-hidden="true"
                          class="dropdown icon"
                        />
                        <div
                          class="menu transition"
                        >
                          <div
                            aria-checked="false"
                            aria-selected="true"
                            class="selected item"
                            role="option"
                            style="pointer-events: all;"
                          >
                            <img
                              class="ui avatar image"
                            />
                            <span
                              class="text"
                            >
                              Sarah Edo
                            </span>
                          </div>
                          <div
                            aria-checked="false"
                            aria-selected="false"
                            class="item"
                            role="option"
                            style="pointer-events: all;"
                          >
                            <img
                              class="ui avatar image"
                            />
                            <span
                              class="text"
                            >
                              Tyler McGinnis
                            </span>
                          </div>
                          <div
                            aria-checked="false"
                            aria-selected="false"
                            class="item"
                            role="option"
                            style="pointer-events: all;"
                          >
                            <img
                              class="ui avatar image"
                            />
                            <span
                              class="text"
                            >
                              Mike Tsamis
                            </span>
                          </div>
                          <div
                            aria-checked="false"
                            aria-selected="false"
                            class="item"
                            role="option"
                            style="pointer-events: all;"
                          >
                            <img
                              class="ui avatar image"
                            />
                            <span
                              class="text"
                            >
                              Zenobia Oshikanlu
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="disabled field"
                    >
                      <button
                        class="ui fluid positive disabled button"
                        data-testid="button"
                        disabled=""
                        tabindex="-1"
                      >
                        Login
                      </button>
                    </div>
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui segments"
    >
      <h4
        class="ui block top attached center aligned header"
      >
        <div
          class="content"
        >
          Employee Polls Web App
        </div>
        <div
          class="sub header"
        >
          Please Sign In to continue
        </div>
      </h4>
      <div>
        <div
          class="ui padded center aligned grid"
        >
          <div
            class="row login"
          >
            <div
              class="sixteen wide column"
            >
              <form
                class="ui form"
              >
                <div
                  data-testid="Test Component"
                >
                  <h2
                    class="ui green header"
                  >
                    Sign In
                  </h2>
                  <div
                    class="required field"
                  >
                    <div
                      aria-expanded="false"
                      class="ui fluid selection scrolling dropdown"
                      data-testid="dropdown"
                      required=""
                      role="listbox"
                      tabindex="0"
                    >
                      <div
                        aria-atomic="true"
                        aria-live="polite"
                        class="divider default text"
                        role="alert"
                      >
                        Select an Employee
                      </div>
                      <i
                        aria-hidden="true"
                        class="dropdown icon"
                      />
                      <div
                        class="menu transition"
                      >
                        <div
                          aria-checked="false"
                          aria-selected="true"
                          class="selected item"
                          role="option"
                          style="pointer-events: all;"
                        >
                          <img
                            class="ui avatar image"
                          />
                          <span
                            class="text"
                          >
                            Sarah Edo
                          </span>
                        </div>
                        <div
                          aria-checked="false"
                          aria-selected="false"
                          class="item"
                          role="option"
                          style="pointer-events: all;"
                        >
                          <img
                            class="ui avatar image"
                          />
                          <span
                            class="text"
                          >
                            Tyler McGinnis
                          </span>
                        </div>
                        <div
                          aria-checked="false"
                          aria-selected="false"
                          class="item"
                          role="option"
                          style="pointer-events: all;"
                        >
                          <img
                            class="ui avatar image"
                          />
                          <span
                            class="text"
                          >
                            Mike Tsamis
                          </span>
                        </div>
                        <div
                          aria-checked="false"
                          aria-selected="false"
                          class="item"
                          role="option"
                          style="pointer-events: all;"
                        >
                          <img
                            class="ui avatar image"
                          />
                          <span
                            class="text"
                          >
                            Zenobia Oshikanlu
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="disabled field"
                  >
                    <button
                      class="ui fluid positive disabled button"
                      data-testid="button"
                      disabled=""
                      tabindex="-1"
                    >
                      Login
                    </button>
                  </div>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
